chall08/edramire.py

```
@@ -0,0 +1,138 @@
+        self.robot = None
+        self.valid = True
+
+    def validate(self, line_str: str):
+        len_line = len(line_str)
+        if len_line < 3 or (self.rows > 0 and len_line != self.cols):
+            return False
+        if any(c not in '+*ME0' for c in line_str):
+            return False
+        e = sum((1 if c == 'E' else 0 for c in line_str), 0)
+        if e > 1 or (e == 1 and self.exit is not None):
+            return False
+        m = sum((1 if c == 'M' else 0 for c in line_str[1:len_line]), 0)
+        if m > 1 or (m == 1 and self.robot is not None):
+            return False
+        if all(c in '+E' for c in line_str):
+            if e == 1 and (line_str.startswith('E') or line_str.endswith('E')):
+                return False
+            if (self.rows > 1
+                    and all(c in '+E' for c in self.lines[self.rows - 1])):
+                return False
+        elif all(c in 'M0*' for c in line_str[1:len_line - 1]):
+            if line_str[0] not in '+E' or line_str[-1] not in '+E':
+                return False
+        else:
+            return False
+        return True
+
+    def is_valid_position(self, row: int, col: int):
+        if (row < 0 or row >= self.rows or col < 0 or col >= self.cols
+                or self.lines[row][col] == '+'):
+            return False
+        return True
+
+    def position_has_bomb(self, row, col):
+        if self.lines[row][col] == '*':
+            return True
+        return False
+
+    def is_exit(self, row, col):
+        if row == self.exit[0] and col == self.exit[1]:
+            return True
+        return False
+
+    def __repr__(self):
+        return f'{self.lines}\n{self.robot}\n{self.exit}'
+
+
+class Robot:
+    moves = {
+        'N': (-1, 0),
+        'E': (0, 1),
+        'S': (1, 0),
+        'W': (0, -1)
+    }
+
+    def __init__(self, row, col):
+        self.row = row
+        self.col = col
+        self.on = False
+        self.destroyed = False
```

You could just raise a custom error on destruction and catch it later !